temporal-test-server/src/main/java/io/temporal/internal/testservice/StateMachines.java

```
@@ -1729,6 +1748,92 @@ private static void heartbeatActivityTask(
     data.heartbeatDetails = details;
   }
 
+  private static void acceptUpdate(
+      RequestContext ctx,
+      UpdateWorkflowExecutionData data,
+      Message msg,
+      long workflowTaskCompletedEventId) {
+    try {
+      Acceptance acceptance = msg.getBody().unpack(Acceptance.class);
+
+      WorkflowExecutionUpdateAcceptedEventAttributes acceptedAttribute =
+          WorkflowExecutionUpdateAcceptedEventAttributes.newBuilder()
+              .setAcceptedRequestSequencingEventId(workflowTaskCompletedEventId - 1)
+              .setProtocolInstanceId(msg.getProtocolInstanceId())
+              .setAcceptedRequestMessageId(acceptance.getAcceptedRequestMessageId())
+              .setAcceptedRequest(acceptance.getAcceptedRequest())
+              .build();
+
+      HistoryEvent event =
+          HistoryEvent.newBuilder()
+              .setEventType(EventType.EVENT_TYPE_WORKFLOW_EXECUTION_UPDATE_ACCEPTED)
+              .setWorkflowExecutionUpdateAcceptedEventAttributes(acceptedAttribute)
+              .build();
+      // If the workflow is finished we can't write more events
+      // to history so if the message was processed after the workflow
+      // was closed there is nothing we can do.
+      // The real server also has this same problem
```

What does the client-side caller of the update see in this situation?

Would love a to see tests for workflow-complete-before-update-accept (probably have to terminate while worker not running) and workflow-complete-before-update-complete (can just hang the update handler and finish the workflow).

---

>What does the client-side caller of the update see in this situation?

This does not impact what the caller side sees. I'll note this shouldn't be an issue once we have message commands.

---

>Would love a to see tests for workflow-complete-before-update-accept (probably have to terminate while worker not running)

I agree but we need the server to support `UPDATE_WORKFLOW_EXECUTION_LIFECYCLE_STAGE_ADMITTED` to write this test.

> workflow-complete-before-update-complete (can just hang the update handler and finish the workflow).

I also agree, I was planning on covering this in the test server and the real server once they get support for `UPDATE_WORKFLOW_EXECUTION_LIFECYCLE_STAGE_ACCEPTED`

---

I wonder if we should just error in workflow-completed situations then instead of silently succeeding (unless I'm misreading)

---

I don't understand why we would error. I am emulating the real server and the real server does not error. Erroring would also not help us?

---

I thought the real server just improperly handled this. Surely an update will error when workflow closed while processing.  I was just saying error here as not-yet-properly implemented on real server so intentionally isn't here either until we see the shape that error takes.